source/data/shader/ShaderLoader.cpp

```diff
@@ -91,55 +107,67 @@ constexpr static GLenum getGlShaderType(const Shader::Type type) {
   return 0;
 }
 
-unsigned int ShaderLoader::loadShader(const std::string &path, Shader::Type type) {
-  std::ifstream file(path);
-  if (not file.is_open()) {
-    Util::Log(Error, "ShaderLoader") << "Could not find shader file " << path;
-  }
-  std::string str;
-  std::string file_contents;
-  while (std::getline(file, str)) {
-    file_contents += str;
-    file_contents.push_back('\n');
-  }
-  const char* buffer = file_contents.c_str();
-
-  GLuint shader = glCreateShader(getGlShaderType(type));
-  glShaderSource(shader, 1, &buffer, NULL);
-  glCompileShader(shader);
-
-  //Error checking
-  GLint success = 0;
-  glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
-
-  const char *shaderType = "<unknown shader type>";
+constexpr const char *ShaderLoader::getShaderTypeString(const Shader::Type type) {
   switch (type) {
   case Shader::Vertex:
-    shaderType = "vertex";
-    break;
+    return "vertex";
   case Shader::Fragment:
-    shaderType = "fragment";
-    break;
+    return "fragment";
   case Shader::Geometry:
-    shaderType = "geometry";
-    break;
+    return "geometry";
   }
+  return "<unknown shader type>";
+}
+
+void ShaderLoader::checkShader(unsigned int shader, const std::string &path, Shader::Type type) {
+  // Error checking
+  GLint success = 0;
+  glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
+
+  const char *shaderType = getShaderTypeString(type);
+
   if (success == GL_TRUE) {
-    Util::Log(Debug, "ShaderLoader") << path << ": " << shaderType << " shader compiled";
+    Util::Log(Debug, "ShaderLoader")
+        << path << ": " << shaderType << " shader compiled";
   } else {
     GLint logSize = 0;
     glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &logSize);
     if (logSize > MaxGlLogSize) {
-      throw std::runtime_error("GL reported compilation log size exceeding maximum");
+      throw std::runtime_error(
+          "GL reported compilation log size exceeding maximum");
     }
     std::unique_ptr<char[]> log(new char[static_cast<unsigned long>(logSize)]);
     glGetShaderInfoLog(shader, logSize, &logSize, log.get());
 
-    Util::Log(Error, "ShaderLoader") << path << ": " << shaderType << " shader compilation "
-      "failed:\n" << log.get();
+    Util::Log(Error, "ShaderLoader")
+        << path << ": " << shaderType << " shader compilation "
+                                         "failed:\n"
+        << log.get();
 
     glDeleteShader(shader);
   }
+}
+
+unsigned int ShaderLoader::loadShader(const std::string &path,
+                                      Shader::Type type) {
+  std::ifstream file(path);
+  if (not file.is_open()) {
+    Util::Log(Error, "ShaderLoader") << "Could not find shader file " << path;
+  }
+  std::string str;
+  std::string file_contents;
```

Comment:
Actually the even better thing to do would be to have a `std::unique_ptr<char[]>` to read the file's contents to directly, instead of using `std::getline` and `file_contents.c_str()`.